### ISSUE TITLE
New compiler: support implicit cast of `char[n]` array to `const string` in function calls

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2175]
CC @fernewelten 

Following script was working in an old compiler:
```
char c[100];

Display(c);
String.Format("%s", c);
```
As far as I understand, compiler would convert char array to a `const char*` pointer there (as if it were an old-style `string`).

The new compiler throws an error on either of these: `Cannot access this array as a whole (did you forget to add "[0]"?`)`

I suggest to support this kind of implicit conversion again, unless there's a good argument against this.

EDIT: actually, this may be considered a bug... but i am not fully certain.

**Comment by ericoporto** [comment]
This looks ambiguous in an utf-8 world, do you expect each char to be a char in utf-8 or a byte?

**Comment by ivan-mogilko** [comment]
> This looks ambiguous in an utf-8 world, do you expect each char to be a char in utf-8 or a byte?

It's not ambiguous, `char` is 8-bit in AGS script.

**Comment by ericoporto** [comment]
For some reason I additionally thought we were using the string class inside to avoid calculating strlen many times but apparently we still don't. :/

**Comment by ivan-mogilko** [comment]
> Btw, this looks more a feature than a bug.

It was working in the old compiler, and syntactically array of chars (bytes) should be castable to `const char*`, so in the end I see this as a regression.

**Comment by fernewelten** [comment]
I agree, this ought to work. I'm looking into what needs to be done to make it happen.

**Comment by fernewelten** [comment]
So here's the problem: A character array doesn't have a guaranteed terminating `'\0'`. 

The compiler can be made to find the bytes that the character array contains and then treat those bytes as a `string`. But when the array is filled to the brim and the resulting `string` is handed on elsewhere, then a buffer overrun will occur.
```
function game_start
{
    char buf[] = new char[5];
    for (int idx = 0; idx < 5; idx++)
        buf[idx] = 'A';
    Display(buf); // ← buffer overrun
}
```

Now if `buf` is a **managed** array (`char buf[]`) instead of a ‘classic’ array (`char buf[5];`) we do know its length at runtime and so we know the maximal number of characters it can contain. So in principle, the compiler could be made to copy the characters to another buffer that is 1 place longer ands ends on a guaranteed `'\0'`, using lots of bytecode. That will turn out to be very ugly and inefficient.

In my opinion, a more feasible solution would be to add a function to the Engine that would take a `char[]` and return a `String`. Perhaps this function can be a secret function that only the compiler knows about, and when the compiler finds code such as `Display(buf)`, above, it will first invoke the secret function and then hand over the resulting `String` to `Display()`. The resulting String would be garbage collected after `Display()` has returned.

For a solution that works with **‘classic’** arrays, the compiler could allocate 1 byte more than specified for all `char` arrays and so force a `'\0'` after such buffers that can't be touched and so can't be overwritten.  

What do you think?

**Comment by ivan-mogilko** [comment]
I dont understand, why the contents of the buffer is a compiler's concern at all?

The original compiler simply passed a pointer to `char array` as a `const char*`, without copying anything or checking its contents.

**Comment by ivan-mogilko** [comment]
So, if I understand correctly, you would like to make this operation *secure*, unlike the old compiler that did not care about this.

I never had such intent when writing this ticket.

There seem to be 2 approaches to this:
1. Invent a method that would make such cast operation safer, possibly at a price of making it bit slower.
2. Add a new compiler option that allows "unsafe" operations at user's responsibility. Such approach is ideal in case this compiler gets backported to ags3, for instance.

In regards to the 1st method. Copying buffer contents or creating a dynamic String is going to be a way too much overhead for this: previously this was just a pointer "cast", now it will become a memory allocation and copy operation. So I would advise against doing so.

If it were possible to pass the buffer's length along, whether acquiring it at compile time or runtime (for dynamic arrays), that might work better. But the question here is: *where* to pass this to. String functions would require a place for this extra argument. And having it as an explicit argument would break API, not to mention that it will be redundant in case of using Strings or string literals. So it would have to be some kind of a secret implicit argument that only engine knows about. Which in turn might make it incompatible with running the previously compiled scripts (so it closes any opportunity to e.g. port the new compiler to backwards compatible engine branch).

Alright, I will give it some more thought when I get more spare time...

**Comment by ivan-mogilko** [comment]
> Now we do know the buffer length at runtime and so we know the maximal number of characters it can contain. So in principle, the compiler could be made to copy the characters to another buffer that is 1 place longer ands ends on a guaranteed '\0', using lots of bytecode. That will turn out to be very ugly and inefficient, but the result can be treated as a proper string.

BTW here's a similar "dumb" solution: compiler could save the last buffer char in register, insert 0 at the end of the buffer before passing to engine function, and put recorded char back after function returned.

**Comment by ivan-mogilko** [comment]
> As an alternative solution the compiler could allocate 1 byte more than specified for all char arrays and so force a '\0' after such buffers that can't be touched and so can't be overwritten. So therefore any such buffer can be treated as string. This would also need special logic in the case of dynamic char arrays for the AGS expression buf.Length, which would have to differentiate between dynamic char arrays and other dynamic arrays.

You know, actually, this sounds like a most optimal solution from everything above. That will cost 1 extra byte per a `char` array (or perhaps old-style `string` too, would the compiler be ported to ags3 and get these supported again), which seems negligible.

A Length of dynamic arrays may even be left as-is, as this byte is not going to be used anywhere, I suppose?
Normally a array class must have 2 properties: length in bytes and length in elements. Dynamic array also has "number of elements", and this is what Length returns? In any case, this seems like something that may be worked around.

**Comment by ivan-mogilko** [comment]
Ah. No, adding extra byte won't be a reliable solution, because when char array is in a struct, that would shift all the struct member offsets. Besides obviously making it more difficult for compiler to calculate offsets, this will cause mismatches with any structs exposed by the engine.

**Comment by fernewelten** [comment]
Here's a point that I'm just realising. There are **two** sides to this issue:
- The old compiler treats a ‘classic’ array as a string when a string is requested
- The old compiler decays _any_ classic arrays (even, e.g., `float f[30];`) into pointers to their first elements when they are used as parameters. There's no way to declare a named parameter that will accept such a ‘decayed’ array, but a ‘decayed’ array can be passed to a variadic function, so plugins might make use of that. 

This means that the new compiler must be able to grok ‘decayed’ arrays as function parameters irrespective of the way it treats `char` arrays when strings are requested.

**Comment by fernewelten** [comment]
There's an actual use case where one might want to do character processing in a `char[]` array and only convert to `String` or `string` at the end:

Let's say we code a word puzzle solver. We will modify single letters in our word thousands of times until we have found our solution. 

If we store the word in a `String`, which is unmodifiable, then each and every modification will lead to a new String and discarding the old String. This will strain the allocation and garbage collection mechanism a _lot_.

It would be preferable in such a situation to keep the letters in an array and to modify array elements until the array is stable. We would like to only convert the array to a `string` or `String` when it needs to be printed on screen.

**Comment by ericoporto** [comment]
If you use a `for` with ‘AppendChar` to reconstruct the String it won't be slow in any use case, I used to use a static char array to store strings in managed structs and didn't knew about this and this was never an issue - drawing always beat anything else in being slow.

I also guess in post utf-8 no one is going to store significant amount of text in this way since then you need an AGS Script library that knows utf-8.

**Comment by ivan-mogilko** [comment]
I am standing for consistency and strict rules in the script. For that reason I believe that if one type of arrays may be treated in certain way by compiler, then others should too. Or none should.
Either we keep support array decay, or not. Or, we support it under a compiler option, which user must enable with full understanding, taking responsibility for the consequence.

Casting char array to a string pointer is a specific thing, it requires a implicit cast allowance between 2 types, but also array decay as a base, if I understand correctly, so I suppose it should not work if "array decay" is not enabled.

But then, char arrays are not secure for use as null-terminated strings, since null terminator is not guaranteed by the system. So if user uses them as such, it should be user's responsibility.

If you look at the other "secure" languages such as C#, they don't let a byte array of unknown length or contents passed as a String parameter. Such char array has to be explicitly converted to String first. But "explicitly" is a key word here. I don't think it's fair to do this secretly, creating an implicit overhead that user is not aware about (and that would be a huge overhead to make a full string copy for example, compared to just passing a pointer or reference of array).

To summarize: either we allow an implicit and unsafe `char[]` to `const string` cast for backwards compatibility, or not, or have a compiler's option that toggles that behavior.

Use existing cases where unsafe arrays are still useful is a separate problem here IMO, it's rather a case of *missing secure alternatives*.
For example:
- We may add mutable String type, something akin to StringBuilder in C# (name does not have to be the same, it could be StringBuffer, CharBuffer, CharArray etc).
- We may add factory methods to String class, which accept array of char and its length, or dynamic array of chars which length may be read at runtime. This will let to explicitly convert a char array to a secure String object before passing further.

**Comment by ericoporto** [comment]
> We may add mutable String type, something akin to StringBuilder in C# (name does not have to be the same, it could be StringBuffer, CharBuffer, CharArray etc)

This looks relatively easy to do as C# has essentially already designed the API - of course we would need to adapt it. But it looks it would have both a Capacity and Length property, a few different Append methods to actually build the string, some method to retrieve the built string, and some way to clear it. But this could be made it's own issue to nail down the essential methods to this thing.

**Comment by ivan-mogilko** [comment]
The regression reported in this ticket has been fixed by #2583
